File: PolicyAnalyzer/SharedCode/AdvAuditSubcategoryItem.cs
using System;
using System.Xml;

namespace SharedCode
{
	// Token: 0x02000046 RID: 70
	internal class AdvAuditSubcategoryItem : BaselineItem
	{
		// Token: 0x060000DF RID: 223 RVA: 0x0000B1F4 File Offset: 0x000093F4
		public AdvAuditSubcategoryItem(XmlNode node)
		{
			base.ReadStdElems(node);
			this.m_sGUID = node["GUID"].InnerText.Trim();
			this.m_sName = node["Name"].InnerText.Trim();
			this.m_sBaselineValue = node["Setting"].InnerText.Trim();
		}

		// Token: 0x060000E0 RID: 224 RVA: 0x00002E80 File Offset: 0x00001080
		public override string SortKey()
		{
			return (base.GetType().Name + "!" + this.m_sGUID).ToLower();
		}

		// Token: 0x060000E1 RID: 225 RVA: 0x00002EA2 File Offset: 0x000010A2
		public override void Evaluate(EffectiveStateSources sources, XmlDocument xOutputDoc)
		{
			this.RetrieveAndCompare(sources);
			base.ImplementIgnoreAlwaysPass();
			this.WriteAdvAuditSubcatResults(xOutputDoc);
		}

		// Token: 0x060000E2 RID: 226 RVA: 0x0000B278 File Offset: 0x00009478
		private XmlNode GetNodeFromSeceditAndAuditpolState(EffectiveStateSources sources)
		{
			if (sources.xmlSeceditAndAuditpolState == null)
			{
				return null;
			}
			string text = string.Concat(new string[]
			{
				"//AuditSubcategory[GUID='",
				this.m_sGUID.ToLower(),
				"' or GUID='",
				this.m_sGUID.ToUpper(),
				"']"
			});
			return sources.xmlSeceditAndAuditpolState.SelectSingleNode(text);
		}

		// Token: 0x060000E3 RID: 227 RVA: 0x0000B2DC File Offset: 0x000094DC
		private void RetrieveAndCompare(EffectiveStateSources sources)
		{
			this.m_bValueFound = false;
			this.m_bEvalPassed = false;
			this.m_bInvalidComparison = false;
			XmlNode nodeFromSeceditAndAuditpolState = this.GetNodeFromSeceditAndAuditpolState(sources);
			if (nodeFromSeceditAndAuditpolState == null)
			{
				return;
			}
			AdvAuditSubcategoryItem advAuditSubcategoryItem = new AdvAuditSubcategoryItem(nodeFromSeceditAndAuditpolState);
			this.m_bValueFound = true;
			this.m_sObservedValue = advAuditSubcategoryItem.m_sBaselineValue;
			this.m_sObservedPolicyName = advAuditSubcategoryItem.m_sGpoName;
			uint num;
			uint num2;
			if (!uint.TryParse(this.m_sBaselineValue, out num) || !uint.TryParse(this.m_sObservedValue, out num2))
			{
				this.m_bInvalidComparison = true;
				return;
			}
			string sEvalOperator = this.m_sEvalOperator;
			uint num3 = <PrivateImplementationDetails>.ComputeStringHash(sEvalOperator);
			if (num3 <= 1519117449U)
			{
				if (num3 != 24087439U)
				{
					if (num3 != 685255070U)
					{
						if (num3 == 1519117449U)
						{
							if (sEvalOperator == "GreaterThanOrEqual")
							{
								this.m_bEvalPassed = (num & num2) == num;
								return;
							}
						}
					}
					else if (sEvalOperator == "NotEqual")
					{
						this.m_bEvalPassed = num != num2;
						return;
					}
				}
				else if (sEvalOperator == "Equal")
				{
					this.m_bEvalPassed = num == num2;
					return;
				}
			}
			else if (num3 <= 2543884473U)
			{
				if (num3 != 2007843446U)
				{
					if (num3 == 2543884473U)
					{
						if (sEvalOperator == "NonNull")
						{
							this.m_bEvalPassed = true;
							return;
						}
					}
				}
				else if (sEvalOperator == "LessThanOrEqual")
				{
					this.m_bEvalPassed = (num | num2) == num;
					return;
				}
			}
			else if (num3 != 2735859570U)
			{
				if (num3 == 3902070617U)
				{
					if (sEvalOperator == "OneOfThese")
					{
						foreach (string text in this.m_sEvalValues.ToLower().Split(new char[] { ',' }))
						{
							if (this.m_sObservedValue == text.Trim())
							{
								this.m_bEvalPassed = true;
								return;
							}
						}
						return;
					}
				}
			}
			else if (sEvalOperator == "Range")
			{
				this.m_bInvalidComparison = true;
				return;
			}
			this.m_bInvalidComparison = true;
		}

		// Token: 0x060000E4 RID: 228 RVA: 0x0000B4F0 File Offset: 0x000096F0
		private void WriteAdvAuditSubcatResults(XmlDocument xOutputDoc)
		{
			XmlElement xmlElement = xOutputDoc.CreateElement("AuditSubcategory");
			XmlElement xmlElement2 = xOutputDoc.CreateElement("GUID");
			XmlElement xmlElement3 = xOutputDoc.CreateElement("Name");
			XmlElement xmlElement4 = xOutputDoc.CreateElement("Setting");
			XmlElement xmlElement5 = xOutputDoc.CreateElement("PolicyName");
			xmlElement2.InnerText = this.m_sGUID;
			xmlElement3.InnerText = this.m_sName;
			xmlElement4.InnerText = this.m_sObservedValue;
			xmlElement5.InnerText = this.m_sObservedPolicyName;
			xmlElement.AppendChild(xmlElement2);
			xmlElement.AppendChild(xmlElement3);
			xmlElement.AppendChild(xmlElement4);
			xmlElement.AppendChild(xmlElement5);
			base.CreateResultsElements(xOutputDoc, xmlElement);
			xOutputDoc.DocumentElement.AppendChild(xmlElement);
		}

		// Token: 0x0400068E RID: 1678
		internal string m_sGUID;

		// Token: 0x0400068F RID: 1679
		internal string m_sName;

		// Token: 0x04000690 RID: 1680
		internal string m_sObservedValue = string.Empty;

		// Token: 0x04000691 RID: 1681
		internal string m_sObservedPolicyName = string.Empty;
	}
}

File: Python/shared/adv_audit_subcategory_item.py
# Exact full port of AdvAuditSubcategoryItem.cs (includes full bitmask logic for GreaterThanOrEqual / LessThanOrEqual)

import xml.etree.ElementTree as ET
from .baseline_item import BaselineItem
from .effective_state_sources import EffectiveStateSources

class AdvAuditSubcategoryItem(BaselineItem):
    def __init__(self, node: ET.Element):
        super().__init__()
        self.ReadStdElems(node)
        guid_elem = node.find("GUID")
        name_elem = node.find("Name")
        setting_elem = node.find("Setting")
        self.m_sGUID = guid_elem.text.strip() if guid_elem is not None and guid_elem.text else ""
        self.m_sName = name_elem.text.strip() if name_elem is not None and name_elem.text else ""
        self.m_sBaselineValue = setting_elem.text.strip() if setting_elem is not None and setting_elem.text else ""

        self.m_sObservedValue: str = ""
        self.m_sObservedPolicyName: str = ""

    def SortKey(self) -> str:
        return (self.__class__.__name__ + "!" + self.m_sGUID).lower()

    def Evaluate(self, sources: EffectiveStateSources, x_output_doc: ET.Element):
        self._retrieve_and_compare(sources)
        self.ImplementIgnoreAlwaysPass()
        self._write_results(x_output_doc)

    def _get_node(self, sources: EffectiveStateSources):
        if sources.xmlSeceditAndAuditpolState is None:
            return None
        guid_lower = self.m_sGUID.lower()
        guid_upper = self.m_sGUID.upper()
        xpath = f"//AuditSubcategory[GUID='{guid_lower}' or GUID='{guid_upper}']"
        return sources.xmlSeceditAndAuditpolState.find(xpath)

    def _retrieve_and_compare(self, sources: EffectiveStateSources):
        self.m_bValueFound = False
        self.m_bEvalPassed = False
        self.m_bInvalidComparison = False

        node = self._get_node(sources)
        if node is None:
            return

        observed = AdvAuditSubcategoryItem(node)
        self.m_bValueFound = True
        self.m_sObservedValue = observed.m_sBaselineValue
        self.m_sObservedPolicyName = observed.m_sGpoName

        try:
            baseline_val = int(self.m_sBaselineValue)
            observed_val = int(self.m_sObservedValue)
        except (ValueError, TypeError):
            self.m_bInvalidComparison = True
            return

        op = self.m_sEvalOperator
        if op == "Equal":
            self.m_bEvalPassed = (baseline_val == observed_val)
        elif op == "NotEqual":
            self.m_bEvalPassed = (baseline_val != observed_val)
        elif op == "NonNull":
            self.m_bEvalPassed = True
        elif op == "GreaterThanOrEqual":
            self.m_bEvalPassed = (baseline_val & observed_val) == baseline_val
        elif op == "LessThanOrEqual":
            self.m_bEvalPassed = (baseline_val | observed_val) == baseline_val
        elif op == "Range":
            self.m_bInvalidComparison = True
        elif op == "OneOfThese":
            for val in self.m_sEvalValues.lower().split(","):
                if self.m_sObservedValue == val.strip():
                    self.m_bEvalPassed = True
                    return
        else:
            self.m_bInvalidComparison = True

    def _write_results(self, x_output_doc: ET.Element):
        root = x_output_doc.getroot()
        if root is None:
            root = ET.SubElement(x_output_doc, "root")
        elem = ET.SubElement(root, "AuditSubcategory")
        ET.SubElement(elem, "GUID").text = self.m_sGUID
        ET.SubElement(elem, "Name").text = self.m_sName
        ET.SubElement(elem, "Setting").text = self.m_sObservedValue
        ET.SubElement(elem, "PolicyName").text = self.m_sObservedPolicyName
        self.CreateResultsElements(x_output_doc, elem)
        root.append(elem)